File: nova-llm-agent/src/utils/rag/processPDFs.js
require('dotenv').config();
const fs = require('fs').promises;
const path = require('path');
const pdfParse = require('pdf-parse');
const { OpenAIEmbeddings } = require('@langchain/openai');
const { PineconeStore } = require('@langchain/pinecone');
const { Pinecone } = require('@pinecone-database/pinecone');
const { RecursiveCharacterTextSplitter } = require('langchain/text_splitter');
const { logger } = require('../logger');

class PDFProcessor {
  constructor() {
    this.embeddings = new OpenAIEmbeddings({
      openAIApiKey: process.env.OPENAI_API_KEY,
    });
    this.pineconeClient = new Pinecone({
      apiKey: process.env.PINECONE_API_KEY,
      host: `${process.env.PINECONE_INDEX}-${process.env.PINECONE_ENV}.pinecone.io`
    });
    this.textSplitter = new RecursiveCharacterTextSplitter({
      chunkSize: 1000,
      chunkOverlap: 200,
    });
  }

  async initialize() {
    // No initialization needed for Pinecone v1.x
  }

  async processPDF(filePath) {
    try {
      logger.info(`Processing PDF: ${filePath}`);

      // Read PDF file
      const dataBuffer = await fs.readFile(filePath);
      const pdfData = await pdfParse(dataBuffer);

      // Split text into chunks
      const chunks = await this.textSplitter.splitText(pdfData.text);

      // Create embeddings and store in Pinecone
      const index = this.pineconeClient.index(process.env.PINECONE_INDEX); // <-- lowercase 'i'
      const vectorStore = await PineconeStore.fromTexts(
        chunks,
        chunks.map((_, i) => ({
          source: path.basename(filePath),
          chunk: i
        })),
        this.embeddings,
        { pineconeIndex: index }
      );

      logger.info(`Successfully processed ${filePath}`);
      return true;
    } catch (error) {
      logger.error(`Error processing PDF ${filePath}:`, error);
      throw error;
    }
  }

  async processAllPDFs() {
    try {
      await this.initialize();

      const pdfsDir = path.join(__dirname, '../../data/pdfs');
      const files = await fs.readdir(pdfsDir);

      for (const file of files) {
        if (file.endsWith('.pdf')) {
          const filePath = path.join(pdfsDir, file);
          await this.processPDF(filePath);
        }
      }

      logger.info('All PDFs processed successfully');
    } catch (error) {
      logger.error('Error processing PDFs:', error);
      throw error;
    }
  }
}

// If this file is run directly
if (require.main === module) {
  const processor = new PDFProcessor();
  processor.processAllPDFs()
    .then(() => process.exit(0))
    .catch((error) => {
      logger.error('Failed to process PDFs:', error);
      process.exit(1);
    });
}

module.exports = PDFProcessor;

File: nova-llm-agent/src/utils/rag/process_pdfs.py
import os
from dotenv import load_dotenv
from pathlib import Path
import PyPDF2
from langchain_openai import OpenAIEmbeddings
from langchain_community.vectorstores import Pinecone as LangchainPinecone
from pinecone import Pinecone as PineconeClient
from langchain.text_splitter import RecursiveCharacterTextSplitter
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class PDFProcessor:
    def __init__(self):
        load_dotenv()
        self.embeddings = OpenAIEmbeddings(
            openai_api_key=os.getenv('OPENAI_API_KEY')
        )
        self.text_splitter = RecursiveCharacterTextSplitter(
            chunk_size=1000,
            chunk_overlap=200
        )
        self.pinecone_api_key = os.getenv('PINECONE_API_KEY')
        self.index_name = os.getenv('PINECONE_INDEX')

    def process_pdf(self, file_path):
        try:
            logger.info(f"Processing PDF: {file_path}")

            # Read PDF file
            with open(file_path, 'rb') as file:
                pdf_reader = PyPDF2.PdfReader(file)
                text = ""
                for page in pdf_reader.pages:
                    page_text = page.extract_text()
                    if page_text:
                        text += page_text

            # Split text into chunks
            chunks = self.text_splitter.split_text(text)
            logger.info(f"Split into {len(chunks)} chunks.")

            # Make metadata for each chunk
            metadatas = [{
                'source': Path(file_path).name,
                'chunk': i
            } for i in range(len(chunks))]

            # Actually upsert vectors to Pinecone
            vectorstore = LangchainPinecone.from_texts(
                texts=chunks,
                embedding=self.embeddings,
                index_name=self.index_name,
                metadatas=metadatas,
            )
            logger.info(f"Uploaded {len(chunks)} vectors to Pinecone index '{self.index_name}'.")

            # Optionally: check how many vectors are in the index
            pc = PineconeClient(api_key=self.pinecone_api_key)
            index = pc.Index(self.index_name)
            stats = index.describe_index_stats()
            logger.info(f"Index stats: {stats}")

            logger.info(f"Successfully processed {file_path}")
            return True

        except Exception as error:
            logger.error(f"Error processing PDF {file_path}: {error}")
            raise error

    def process_all_pdfs(self):
        try:
            pdfs_dir = (Path(__file__).parent.parent.parent / 'data' / 'pdfs').resolve()
            pdf_files = list(pdfs_dir.glob('*.pdf'))
            if not pdf_files:
                logger.warning(f"No PDF files found in {pdfs_dir}")
            for file in pdf_files:
                self.process_pdf(str(file))

            logger.info("All PDFs processed successfully")

        except Exception as error:
            logger.error(f"Error processing PDFs: {error}")
            raise error

if __name__ == "__main__":
    processor = PDFProcessor()
    processor.process_all_pdfs()
